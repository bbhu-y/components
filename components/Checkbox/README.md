---
category: Components
title: Checkbox
subtitle: 多选框
---

复选框控件 - 对 ant-design 的 Checkbox 组件的包装


## API
### Checkbox
| 名称 | 类型 | 默认值 | 描述 |
|:-----|:-----|:-----|:-----|
| checked | bool | false |  是否选中 |
| disabled | bool | false |  是否禁用 |
| beforeClick | function |  |  点击改变状态前的判断函数，返回true则点击有效，返回false则点击无效，支持返回 Promise 包裹值<br><br>**Signature:**<br>`function() => undefined`<br> |
| onChange | function |  |  变化时回调函数<br><br>**Signature:**<br>`function(value: boolean) => void`<br>*value:* null |

### VCheckbox
| 名称 | 类型 | 默认值 | 描述 |
|:-----|:-----|:-----|:-----|
| children | node | null |  组件内容 |
| className | string | null |  类名 |
| title | string | null |  标题 |
| hidden | bool | false |  是否隐藏 |
| inline | bool | false |  是否是inline-block模式 |
