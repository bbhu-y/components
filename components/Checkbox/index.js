import React from 'react'
import PropTypes from 'prop-types'
import { Checkbox } from 'antd'

class MyCheckbox extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      checked: this.props.checked || false
    }
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.checked !== this.props.checked ||
      nextProps.checked !== this.state.checked) {
      this.setState({
        checked: nextProps.checked
      })
    }
  }
  handleChange = (event) => {
    const { beforeClick } = this.props
    Promise.resolve().then(() => {
      if (beforeClick) {
        return beforeClick(this.state.checked, event)
      } else {
        return true
      }
    }).then((clickAble) => {
      if (clickAble) {
        let checked = !this.state.checked
        this.setState({checked}, () => {
          this.props.onChange && this.props.onChange(checked, event)
        })
      }
    })
  }
  handlClick = (e) => {
    if (this.props.stopClickPropagation) {
      e.stopPropagation()
    }
  }
  render() {
    let checked = this.state.checked
    return (
      <Checkbox
        className={'Checkbox ' + (this.props.className || '')}
        checked={checked}
        indeterminate={!checked && this.props.indeterminate}
        onChange={this.handleChange}
        onClick={this.handlClick}
        disabled={this.props.disabled}
      >
        {this.props.children}
      </Checkbox>
    )
  }
}

MyCheckbox.propTypes = {
  /**
   * 是否选中
   */
  checked: PropTypes.bool,
  /**
   * 是否部分选中，只在未选中时控制样式用
   */
  indeterminate: PropTypes.bool,
  /**
   * 是否阻止点击事件冒泡
   */
  stopClickPropagation: PropTypes.bool,
  /**
   * 是否禁用
   */
  disabled: PropTypes.bool,
  /**
   * 点击改变状态前的判断函数，返回true则点击有效，返回false则点击无效，支持返回 Promise 包裹值
   * @returns {boolean|Promise}
   */
  beforeClick: PropTypes.func,
  /**
   * 变化时回调函数
   * @param {boolean} value
   */
  onChange: PropTypes.func
}

MyCheckbox.defaultProps = {
  checked: false,
  disabled: false,
  stopClickPropagation: true
}

export default MyCheckbox