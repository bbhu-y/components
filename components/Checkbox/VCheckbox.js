import React from 'react'
import PropTypes from 'prop-types'
import { propTypes, withFormsy } from 'formsy-react'
import ControlGroup from '../ControlGroup'
import FormsyError from '../FormsyError'
import Checkbox from '../Checkbox'

class VCheckbox extends React.Component {
  changeValue = (value) => {
    this.props.setValue(value);
    this.props.onChange && this.props.onChange(value)
  }
  render() {
    let { className, title, hidden, inline } = this.props
    const groupClassName = `${className || ''} VCheckbox`

    return (
      <ControlGroup
        className={groupClassName}
        title={title}
        hidden={hidden}
        inline={inline}
      >
        <Checkbox
          onChange={this.changeValue}
          checked={this.props.getValue()}
          disabled={this.props.disabled}
          beforeClick={this.props.beforeClick}
        >
          {this.props.children}
        </Checkbox>
        <FormsyError {...this.props}/>
      </ControlGroup>
    );
  }
}

VCheckbox.propTypes = {
  /**
   * 组件内容
   */
  children: PropTypes.node,
  /**
   * 类名
   */
  className: PropTypes.string,
  /**
   * 标题
   */
  title: PropTypes.string,
  /**
   * 是否隐藏
   */
  hidden: PropTypes.bool,
  /**
   * 是否是inline-block模式
   */
  inline: PropTypes.bool,

  ...propTypes,
}

VCheckbox.defaultProps = {
  children: null,
  className: null,
  title: null,
  hidden: false,
  inline: false
}

export default withFormsy(VCheckbox)