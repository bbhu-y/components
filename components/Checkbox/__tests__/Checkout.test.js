import React from 'react'
import { mount, shallow } from 'enzyme'
import { sleep } from '../../../jest.utils'
import toJson from 'enzyme-to-json'
import Checkbox from '../index'

describe('Checkbox', () => {
  it('render basically', () => {
    const wrapper = shallow(<Checkbox checked={true}>checked</Checkbox>);

    expect(toJson(wrapper.render())).toMatchSnapshot()

    expect(wrapper.contains('checked')).toEqual(true)
  })

  it('onChange', async () => {
    const testState = { checked: true }
    const onChange = () => {
      testState.checked = !testState.checked
    }
    const mountWrapper = mount(<Checkbox checked={testState.checked} onChange={onChange} beforeClick={() => true}>checked</Checkbox>)

    expect(
      mountWrapper
        .find('.ant-checkbox-input')
        .at(0)
        .prop('checked'),
    ).toBe(true)

    mountWrapper
      .find('.ant-checkbox-input')
      .at(0)
      .simulate('change')

    await sleep(100)

    expect(testState.checked).toEqual(false)
  })

  it('beforeClick', async () => {
    const testState = { checked: true }
    const onChange = () => {
      testState.checked = !testState.checked
    }
    const mountWrapper = mount(<Checkbox checked={true} onChange={onChange} beforeClick={() => false}/>)
    mountWrapper
      .find('.ant-checkbox-input')
      .at(0)
      .simulate('change')

    await sleep(100)

    expect(testState.checked).toEqual(true)
  })

  it('disabled and indeterminate', () => {
    const wrapper = mount(<Checkbox checked={false} disabled indeterminate/>)

    expect(toJson(wrapper.render())).toMatchSnapshot()

    expect(wrapper.find('.ant-checkbox-input').at(0).prop('disabled')).toEqual(true)
    expect(wrapper.exists('.ant-checkbox-indeterminate')).toEqual(true)
  })

  it('stopClickPropagation', () => {
    const onClick = jest.fn()
    const wrapper = mount(<div onClick={onClick}><Checkbox checked={false} stopClickPropagation/></div>)
    wrapper.find('.ant-checkbox-input').at(0).simulate('click')
    expect(onClick).not.toHaveBeenCalled()
  })
})