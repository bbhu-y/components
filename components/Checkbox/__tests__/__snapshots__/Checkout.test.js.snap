// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox disabled and indeterminate 1`] = `
<label
  class="Checkbox  ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
>
  <span
    class="ant-checkbox ant-checkbox-indeterminate ant-checkbox-disabled"
  >
    <input
      class="ant-checkbox-input"
      disabled=""
      type="checkbox"
      value=""
    />
    <span
      class="ant-checkbox-inner"
    />
  </span>
</label>
`;

exports[`Checkbox render basically 1`] = `
<label
  class="Checkbox  ant-checkbox-wrapper ant-checkbox-wrapper-checked"
>
  <span
    class="ant-checkbox ant-checkbox-checked"
  >
    <input
      checked=""
      class="ant-checkbox-input"
      type="checkbox"
    />
    <span
      class="ant-checkbox-inner"
    />
  </span>
  <span>
    checked
  </span>
</label>
`;
